File: 2022/18/18.py
# taken from https://github.com/juanplopes/advent-of-code-2022/blob/main/day18.py

import sys, collections

input = open("input").read()
cubes = set(tuple(map(int, x.split(","))) for x in input.splitlines())
mins, maxs = [min(x) - 1 for x in zip(*cubes)], [max(x) + 1 for x in zip(*cubes)]


def neighbors(x, y, z):
    yield from ((x + 1, y, z), (x, y + 1, z), (x, y, z + 1))
    yield from ((x - 1, y, z), (x, y - 1, z), (x, y, z - 1))


total1 = sum(n not in cubes for c in cubes for n in neighbors(*c))
total2 = 0

# Create queue of cubes which starts with the previously found min
# As well, consider all known cubes as 'visited' but some cubes could have empty space
# and we don't want to visit twice.
Q, visited = collections.deque([tuple(mins)]), set(cubes)

# while we have gas in the tank
while len(Q):
    # for each neighboring cube of a popped cube from the deque
    for cube in neighbors(*Q.popleft()):

        # Check if that coordinate is within the bounds of the seen min/max
        # if it isn't, then we skip this neighboring cube.
        if not all(a <= x <= b for a, b, x in zip(mins, maxs, cube)):
            continue

        # We've arrived here, so we know that the cube is within bounds
        # add if the found cube is in the cubes set
        total2 += cube in cubes

        # If we've already visited, then continue
        if cube in visited:
            continue

        # Otherwise add to the visited set, and append as a future cube
        # . to visit
        visited.add(cube)
        Q.append(cube)
print(total1, total2)

File: 2022/18/18.go
package main

import (
	"bufio"
	"io"
	"strconv"
	"strings"
)

type axis int

const (
	X axis = iota
	Y
	Z
)

type cube [3]int

func (c *cube) neighbors() []cube {
	cubes := make([]cube, 0, 6)

	for _, dx := range []int{-1, 1} {
		cubes = append(cubes, cube{c[X] + dx, c[Y], c[Z]})
	}

	for _, dy := range []int{-1, 1} {
		cubes = append(cubes, cube{c[X], c[Y] + dy, c[Z]})
	}

	for _, dz := range []int{-1, 1} {
		cubes = append(cubes, cube{c[X], c[Y], c[Z] + dz})
	}

	return cubes
}

func readInput(input io.Reader) (cubes []cube) {

	scanner := bufio.NewScanner(input)

	for scanner.Scan() {
		line := scanner.Text()

		split := strings.Split(line, ",")

		x, _ := strconv.Atoi(split[X])
		y, _ := strconv.Atoi(split[Y])
		z, _ := strconv.Atoi(split[Z])

		cubes = append(cubes, cube{x, y, z})
	}

	return cubes

}

func run(input io.Reader) (int, int) {
	cubes := readInput(input)

	cubeMap := make(map[cube]bool, len(cubes))

	min, max := cube{100, 100, 100}, cube{-100, -100, -100}
	for _, cube := range cubes {
		cubeMap[cube] = true

		for _, D := range []axis{X, Y, Z} {
			if cube[D] < min[D] {
				min[D] = cube[D]
			}

			if cube[D] > max[D] {
				max[D] = cube[D]
			}
		}
	}

	for _, D := range []axis{X, Y, Z} {
		min[D] -= 1
		max[D] += 1
	}

	c1 := 0
	for _, cube := range cubes {
		for _, n := range cube.neighbors() {
			if _, ok := cubeMap[n]; !ok {
				c1++
			}
		}
	}

	queue := []cube{min}
	visited := map[cube]bool{}
	for k, v := range cubeMap {
		visited[k] = v
	}
	c2 := 0

	for len(queue) > 0 {
		cur := queue[0]
		queue = queue[1:]

		for _, n := range cur.neighbors() {
			if !(n[X] >= min[X] && n[X] <= max[X] && n[Y] >= min[Y] && n[Y] <= max[Y] && n[Z] >= min[Z] && n[Z] <= max[Z]) {
				continue
			}

			if _, ok := cubeMap[n]; ok {
				c2++
			}

			if _, visited := visited[n]; visited {
				continue
			}

			visited[n] = true
			queue = append(queue, n)
		}
	}

	return c1, c2
}
